Mathlib/Order/Preorder/Finite.lean

```diff
@@ -0,0 +1,137 @@
+lemma exists_minimal (hs : s.Nonempty) : ∃ i, Minimal (· ∈ s) i := s.exists_minimalFor id hs
+
+lemma exists_le_maximal (s : Finset α) (ha : a ∈ s) : ∃ b, a ≤ b ∧ Maximal (· ∈ s) b := by
+  classical
+  obtain ⟨b, hb, hab, hbmin⟩ : ∃ b ∈ s, a ≤ b ∧ _ := by
+    simpa [Maximal, and_assoc] using {x ∈ s | a ≤ x}.exists_maximal ⟨a, mem_filter.2 ⟨ha, le_rfl⟩⟩
+  exact ⟨b, hab, hb, fun c hc hbc ↦ hbmin hc (hab.trans hbc) hbc⟩
+
+lemma exists_le_minimal (s : Finset α) (ha : a ∈ s) : ∃ b ≤ a, Minimal (· ∈ s) b :=
+  exists_le_maximal (α := αᵒᵈ) s ha
+
+@[deprecated (since := "2025-05-04")] alias exists_minimal_le := exists_le_minimal
+
+end Finset
+
+namespace Set
+section Preorder
+variable [Preorder α] {s : Set α} {a : α}
+
+lemma Finite.exists_maximalFor (f : ι → α) (s : Set ι) (h : s.Finite) (hs : s.Nonempty) :
+    ∃ i, MaximalFor (· ∈ s) f i := by
+  lift s to Finset ι using h; exact s.exists_maximalFor f hs
+
+lemma Finite.exists_minimalFor (f : ι → α) (s : Set ι) (h : s.Finite) (hs : s.Nonempty) :
+    ∃ i, MinimalFor (· ∈ s) f i := Finite.exists_maximalFor (α := αᵒᵈ) f s h hs
+
+lemma Finite.exists_maximal (h : s.Finite) (hs : s.Nonempty) : ∃ i, Maximal (· ∈ s) i :=
+  h.exists_maximalFor id _ hs
+
+lemma Finite.exists_minimal (h : s.Finite) (hs : s.Nonempty) : ∃ i, Minimal (· ∈ s) i :=
+  h.exists_minimalFor id _ hs
+
+/-- A version of `Finite.exists_maximalFor` with the (weaker) hypothesis that the image of `s`
+is finite rather than `s` itself. -/
+lemma Finite.exists_maximalFor' (f : ι → α) (s : Set ι) (h : (f '' s).Finite) (hs : s.Nonempty) :
+    ∃ i, MaximalFor (· ∈ s) f i := by
+  obtain ⟨_, ⟨a, ha, rfl⟩, hmax⟩ := Finite.exists_maximalFor id (f '' s) h (hs.image f)
+  exact ⟨a, ha, fun a' ha' hf ↦ hmax (mem_image_of_mem f ha') hf⟩
+
+/-- A version of `Finite.exists_minimalFor` with the (weaker) hypothesis that the image of `s`
+is finite rather than `s` itself. -/
+lemma Finite.exists_minimalFor' (f : ι → α) (s : Set ι) (h : (f '' s).Finite) (hs : s.Nonempty) :
+    ∃ i, MinimalFor (· ∈ s) f i := h.exists_maximalFor' (α := αᵒᵈ) f s hs
+
+@[deprecated (since := "2025-05-04")] alias Finite.exists_maximal_wrt := Finite.exists_maximalFor
+@[deprecated (since := "2025-05-04")] alias Finite.exists_minimal_wrt := Finite.exists_minimalFor
+@[deprecated (since := "2025-05-04")] alias Finite.exists_maximal_wrt' := Finite.exists_maximalFor'
+@[deprecated (since := "2025-05-04")] alias Finite.exists_minimal_wrt' := Finite.exists_minimalFor'
+
+lemma Finite.exists_le_maximal (hs : s.Finite) (ha : a ∈ s) : ∃ b, a ≤ b ∧ Maximal (· ∈ s) b := by
+  lift s to Finset α using hs; exact s.exists_le_maximal ha
+
+lemma Finite.exists_le_minimal (hs : s.Finite) (ha : a ∈ s) : ∃ b, b ≤ a ∧ Minimal (· ∈ s) b := by
+  lift s to Finset α using hs; exact s.exists_le_minimal ha
+
+variable [Nonempty α]
+
+lemma infinite_of_forall_exists_gt (h : ∀ a, ∃ b ∈ s, a < b) : s.Infinite := by
```

Comment:
Either this lemma doesn't belong in the file, or the module doc doesn't make sense!

Reply:
Really? It's the contraposite of a statement about finite sets!

Reply:
...so it's not a statement about finite sets?

Reply:
I strongly believe `P → ¬ Q` and `Q → ¬ P` should be in the same mathlib file

Reply:
Then the module doc needs updating, as I suggested above.